### PR TITLE
[EFM] Refactored encoding and decoding of DKG data during bootstrapping

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/onflow/flow-go/issues/6726

### Context

Refactored encoding structures to correctly represent available data, previously it was missing information about participants. Also grouped it by participant to avoid relying on indexing for 3 structures, now instead of 3 different slices we have 1.

### CODE REVIEW

**Review by jordanschalm — APPROVED** [review]

File: state/protocol/inmem/encodable.go [hunk 1]
```diff
@@ -86,10 +86,17 @@ func (snap EncodableSnapshot) LatestSealedResult() (*flow.ExecutionResult, error
 	return nil, fmt.Errorf("LatestSealedResult: unreachable for correctly formatted sealing segments")
 }
 
+// EncodableFullDKG encodes DKG data for all participants and the group key. Effectively encodes output of the DKG.
 type EncodableFullDKG struct {
```
Comment (AlexHentschel) [inline state/protocol/inmem/encodable.go:90]: DKG stands for _distributed_ key generation. But we only use this the encode the output of a _centralized_ key generation! Furthermore, the struct contains the key set for a threshold signature scheme (which is conceptually unrelated to a DKG)  Tarak has repeatedly reminded us to not conflate the terminology of Threshold Key Cryptography (in our particular case used for the Random Beacon) with DKG. 

As we are touching this code, I would suggest to also clean up the terminology:
* rename `EncodableFullDKG` to `ThresholdKeySet`
* rename `EncodableDKGParticipant` to `ThresholdParticipant`

```suggestion
// ThresholdKeySet contains the key set for a threshold signature scheme. Typically, the ThresholdKeySet is used to
// encode the output of a trusted setup. In general, signature scheme is configured with a threshold parameter t,
// which is the number of malicious colluding nodes the signature scheme is safe against. To balance liveness and
// safety, the Flow protocol fixes threshold to t = floor((n-1)/2), for n the number of parties in the threshold
// cryptography scheme, specifically n = len(Participants).
// Without loss of generality, our threshold cryptography protocol with n parties identifies the individual
// participants by the indices {0, 1, …, n-1}. The slice Participants is ordered accordingly. 
type ThresholdKeySet struct {
```

File: state/protocol/inmem/encodable.go [hunk 2]
```diff
@@ -86,10 +86,17 @@ func (snap EncodableSnapshot) LatestSealedResult() (*flow.ExecutionResult, error
...
-	GroupKey      encodable.RandomBeaconPubKey
-	PrivKeyShares []encodable.RandomBeaconPrivKey
-	PubKeyShares  []encodable.RandomBeaconPubKey
+	GroupKey     encodable.RandomBeaconPubKey
+	Participants []EncodableDKGParticipant
+}
+
+// EncodableDKGParticipant encodes DKG data for single participant.
+type EncodableDKGParticipant struct {
```
Comment (AlexHentschel) [inline state/protocol/inmem/encodable.go:96]:
```suggestion
// ThresholdParticipant encodes the threshold key data for single participant.
type ThresholdParticipant struct {
```

File: cmd/bootstrap/cmd/finalize.go [hunk 3]
```diff
@@ -349,7 +349,7 @@ func readRootBlock() *flow.Block {
 
 // readDKGData reads DKG data from disc, this file needs to be prepared with
 // rootblock command
-func readDKGData() dkg.DKGData {
+func readDKGData() (dkg.DKGData, flow.DKGIndexMap) {
```
Comment (AlexHentschel) [inline cmd/bootstrap/cmd/finalize.go:352]:
```suggestion
// readRandomBeaconKeys reads the threshold key data from disc. 
// This file needs to be prepared with rootblock command
func readRandomBeaconKeys() (dkg.DKGData, flow.DKGIndexMap) {
```

File: cmd/bootstrap/cmd/finalize.go [hunk 4]
```diff
@@ -349,7 +349,7 @@ func readRootBlock() *flow.Block {
...
 	encodableDKG, err := utils.ReadData[inmem.EncodableFullDKG](flagDKGDataPath)
 	if err != nil {
 		log.Fatal().Err(err).Msg("could not read DKG data")
```
Comment (AlexHentschel) [inline cmd/bootstrap/cmd/finalize.go:355]:
```suggestion
		log.Fatal().Err(err).Msg("loading threshold key data for Random Beacon failed")
```

**Review by AlexHentschel — APPROVED** [review]
thank for the cleanup.

**Pull request merged**